File: src/main/java/com/frank/usecase/ShelfUseCase.java
package com.frank.usecase;

import com.frank.entity.book.BookCollection;
import com.frank.entity.book.Book;
import com.frank.entity.book.BookGenre;
import com.frank.entity.book.BookShelf;

public class ShelfUseCase {

    public BookShelf newShelf(BookGenre bookGenre, int placesCount) {
        return new BookShelf(placesCount, bookGenre);
    }

    public BookShelf newShelf(int placesCount) {
        return newShelf(null, placesCount);
    }

    public void addBookToShelf(Book book, BookShelf bookShelf) {
        BookCollection bookCollection = new BookCollection();
        bookCollection.add(book);
        bookShelf.place(bookCollection);
    }

    public void moveCollectionFromShelfToShelf(BookCollection bookCollection, BookShelf src, BookShelf dest) {
        src.remove(bookCollection);
        dest.place(bookCollection);
    }
}

File: src/main/kotlin/com/frank/usecase/ShelfUseCase.kt
package com.frank.usecase

import com.frank.entity.book.BookCollection
import com.frank.entity.book.Book
import com.frank.entity.book.BookGenre
import com.frank.entity.book.BookShelf

class ShelfUseCase {

    fun newShelf(bookGenre: BookGenre?, placesCount: Int): BookShelf {
        return BookShelf(placesCount, bookGenre!!)
    }

    fun newShelf(placesCount: Int): BookShelf {
        // Point 1: not nullable anymore
//        return newShelf(null, placesCount)
        return newShelf(BookGenre(), placesCount)
    }

    fun addBookToShelf(book: Book, bookShelf: BookShelf) {
        val bookCollection = BookCollection()
        bookCollection.add(book)
        bookShelf.place(bookCollection)
    }

    fun moveCollectionFromShelfToShelf(bookCollection: BookCollection, src: BookShelf, dest: BookShelf) {
        src.remove(bookCollection)
        dest.place(bookCollection)
    }
}
